content/blog/2020-11-30-super-optimization/index.md

```diff
@@ -0,0 +1,49 @@
+
+Then super optimization then preform a test of the generated program determine if the generated program is equivalent with the orgnial program and if they are the same, then generated programwill be the optimal porgram for the ISA.  Although this soulds simple, verifing the generated program is a non trival task.  One of the sloution proposed in the paper was boolean test.  However as the complexity of some instructions have expressio of the order of 2^31 minterms (Massalin), it has caused a constriant on the number of instrcitios that can be generated and verified in a reasonable time to be limited to 3 instructions at the time of publication.  Therefore although the boolean test is intuitive, and correct a alternative test will be needed in order to test programs wiht larger amount of instructions.  
+
+The paper also has proposed a probailistict execution test in which the superoptimizer picks a test vectr and see if the output results are the same between the test and generated program.  Although this test is not compprehansive, the author claims that there was never a case where probailistict execution test passes while boolean test fails.  The time it takes for exhaustive search can be futher reduced by ulitization of pruing on the generated sequences that are equivalent to shorter sequences.  
+
+## Results
+
+One of the most intresting part of this paper is the results.  Although compare to other papers where they have a long compression of performance, superoptimizer reported it's result by comparing the execution of the input porgram and assembly.  For example, given the following signum program: 
+<img src="superin.png" style="width: 30%;">
+
+Getting signum program to be immplemented in less then 9 assembly statments is already hard.  But superoptimizer managed to convert the program into a series of adds, subtracts, negation without any branches.  Resulting in the following results:
+<img src="superout.png" style="width: 30%;">
+By removing branches and the number of instructions, this immplemation has provide huge benfits to morden out of order processors (remove the need of branch predictions) and in both number of execution cycles.    Other results are similar interesting and that demestrated clearly the different then what a regular complier would generate.  
```

Comment:
Can you elaborate a little bit on why you're convinced these output programs are unattainable for a conventional compiler? It would be interesting to hear your thoughts on what the limits of the traditional approach are.